### DISCUSSION TITLE
如何全局拦截layui发出的全部ajax请求

### DISCUSSION THREAD

**Discussion by ahhyong** [layui/layui#1482] [问答] [answered]
### 议题条件

- [X] 我确认已查看官方使用文档：https://layui.dev ，但没有找到相关解决方案。
- [X] 我确认已在 [Issues](https://github.com/layui/layui/issues?q=is%3Aissue ) 中搜索过类似的问题，但没有找到相关解决方案。
- [X] 我已仔细阅读： 🍀[ Layui Issue 贡献指南](https://github.com/layui/layui/blob/main/CONTRIBUTING.md)

### 议题类型

功能请求

### 使用版本

v2.8.18

### 问题描述

需要做一些全局异常和响应结果的处理，如何全局拦截layui发出的全部ajax请求？

### 业务代码

```auto
无
```

### 浏览器

全部

### 演示地址

无

### 友好承诺

- [X] 我承诺将本着相互尊重、理解和友善的态度进行交流，共同维护 Layui 良好的社区氛围。

**Comment by Sight-wcg** [comment]
参考 https://api.jquery.com/jquery.ajaxsetup/#jQuery-ajaxSetup-options

**Reply by ahhyong** [reply]
代码里已经添加了jquery全局拦截，自己的ajax请求是可以拦截到的，但是layui的拦截不到

**Reply by Sight-wcg** [reply]
试试 layui.jquery.ajaxSetup()

**Reply by ahhyong** [reply]
试过了，拦截不到

**Reply by Sight-wcg** [reply]
你能提供一个重现吗？

**Reply by ahhyong** [reply]
就是页面js开始的地方添加下面这一段：

layui.jquery.ajaxSetup({
    beforeSend(xhr) {
    },
    complete(xhr, status) {
        console.log("complete=========")
    },
    success(result, status, xhr) {
        console.log("success------------")
    },
    error: function (xhr, status, error) {
        console.log("error+++++++++")
    }
});

**Reply by Sight-wcg** [reply]
![image](https://github.com/layui/layui/assets/26325820/400f7a3e-999b-4619-a090-876b9a30f0c2)

**Reply by ahhyong** [reply]
试了一下您提供的方式，确实可以拦截到layui发出的请求。    但是为啥$.ajaxSetup可以拦截自己发出的ajax请求，但是拦截不到layui的，不知道您可知道原因是什么？

**Reply by Sight-wcg** [reply, marked as answer]
因为 layu.table 内部的一些处理逻辑用到了 ajax 的 success, error 等回调，所以 ajaxSetup 设置的全局回调对其无效。
jQuery 之所以废弃了在 ajaxSetup 的选项中设置全局回调的方法，就是因为目前遇到的问题，缺点太明显。
建议始终按照 jQuery 文档推荐的方法，基于事件设置全局回调。

**Reply by ahhyong** [reply]
感谢您的耐心解答

**Reply by Sight-wcg** [reply]
例如这个问题 ，也是 ajaxSetup 选项设置全局回调的缺点。 https://gitee.com/layui/layui/issues/I8GP73  ,也可以通过事件设置全局回调解决。

**Reply by DevilSquirrel** [reply]
> 试了一下您提供的方式，确实可以拦截到layui发出的请求。 但是为啥$.ajaxSetup可以拦截自己发出的ajax请求，但是拦截不到layui的，不知道您可知道原因是什么？

使layui的dev模块拆分的版本ajaxSetup 是可以拦截的，使用压缩版本不能拦截到。

**Comment by Sight-wcg** [comment]
此问题将被转移到 Discussions

> Layui 的 issue 只受理 「Bug 报告」和「功能请求」。如果是关于如何使用、功能疑惑或其他业务相关的问题建议在 Discussions 寻找社区帮助。

**Reply by sentsim** [reply]
赞，曾经社区有太多 how to 的问题，是时候培养大家的主动思考习惯。
另外，类似的 issue，晨星后续直接先添加 `discussion` 标签即可，github-action bot 会自动回复相关信息。见：
https://github.com/layui/layui/blob/main/.github/workflows/issue-labeled.yml#L59-L70

**Comment by cvpv** [comment]
我是直接自己封装了jquery的ajax
```JavaScript
/**
 * 全站 Ajax 接口请求
 */
function request(option) {
  return new Promise(function (resolve, reject) {
    const { url, method, data } = option
    $.ajax({
      url: url,
      method: method || 'get',
      data: data,
      ...option,
    })
      .done(function (response) {
        let res = response
        if (typeof response === "string") res = JSON.parse(response)
        if (res.result) {
          resolve(res)
        } else {
          if (!option.noMsg && (res.msg || res.message)) {
            msg(res.msg || res.message || '发生了一些错误')
          }
          reject(res.msg || res.message || '服务端未返回约定参数')
        }
      })
      .fail(function (jqXHR, textStatus, errorThrown) {
        console.warn(jqXHR, textStatus, errorThrown)
        reject(textStatus)
      })
  })
}
```

**Reply by ahhyong** [reply]
这是对jquery ajax的封装。自己通过$.ajax发出的请求 我是可以通过$.ajaxSetup拦截到，但是比如layui的table.render发出的请求就拦截不到
